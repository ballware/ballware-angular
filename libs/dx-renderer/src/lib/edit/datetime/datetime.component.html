<ballware-edit-fieldset *ngIf="layoutItem" [layoutItem]="layoutItem">
  <dx-date-box *ngIf="layoutItem && layoutItem.options && layoutItem.options.dataMember"
    [name]="layoutItem.options.dataMember"
    [type]="type"
    [displayFormat]="displayFormat"
    [(value)]="value"
    [readOnly]="(readonly$ | async) ?? false"
    [showClearButton]="true"
    >
      <dx-validator>
        <dxi-validation-rule *ngIf="(required$ | async) ?? false" type="required"></dxi-validation-rule>
      </dx-validator>
  </dx-date-box>
</ballware-edit-fieldset>
