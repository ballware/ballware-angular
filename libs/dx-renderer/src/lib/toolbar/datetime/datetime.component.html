<dx-date-box
    #dateBox
    *ngIf="item"
    [label]="item.caption ?? ''"
    [width]="item.width ?? '220px'"
    type="datetime"
    [displayFormat]="getDisplayFormat()"
    [hint]="item.caption"
    (onValueChanged)="onValueChanged($event)">
</dx-date-box>

