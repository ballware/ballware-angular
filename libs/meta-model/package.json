{
  "name": "@ballware/meta-model",
  "version": "8.0.1",
  "type": "commonjs"
}
