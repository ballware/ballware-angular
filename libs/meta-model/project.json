{
  "name": "meta-model",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/meta-model/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": [
        "{options.outputPath}"
      ],
      "options": {
        "outputPath": "dist/libs/meta-model",
        "main": "libs/meta-model/src/index.ts",
        "tsConfig": "libs/meta-model/tsconfig.lib.json",
        "assets": [
          "libs/meta-model/*.md"
        ]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs meta-model {args.ver} {args.tag}"
      },
      "dependsOn": [
        "build"
      ]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": [
        "{options.outputFile}"
      ],
      "options": {
        "lintFilePatterns": [
          "libs/meta-model/**/*.ts"
        ]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": [
        "{workspaceRoot}/coverage/{projectRoot}"
      ],
      "options": {
        "jestConfig": "libs/meta-model/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
