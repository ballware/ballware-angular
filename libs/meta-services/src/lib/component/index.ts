import { StoreModule } from '@ngrx/store';
import { componentReducer } from './component.reducer';
import { componentFeatureKey } from './component.state';

export * from './component.actions';
export * from './component.state';

export const ComponentFeatureModule = StoreModule.forFeature(componentFeatureKey, componentReducer);