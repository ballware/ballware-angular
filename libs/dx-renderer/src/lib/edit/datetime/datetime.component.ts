import { Component, Inject, Input, OnInit } from '@angular/core';
import { EditLayoutItem } from '@ballware/meta-model';
import { EDIT_SERVICE, EditItemRef, EditService, Translator, TRANSLATOR } from '@ballware/meta-services';
import { DateType } from 'devextreme/ui/date_box';
import { takeUntil } from 'rxjs';
import { WithDestroy } from '../../utils/withdestroy';
import { WithEditItemLifecycle } from '../../utils/withedititemlivecycle';
import { WithReadonly } from '../../utils/withreadonly';
import { WithRequired } from '../../utils/withrequired';
import { WithValidation } from '../../utils/withvalidation';
import { WithValue } from '../../utils/withvalue';
import { WithVisible } from '../../utils/withvisible';
import { DxDateBoxModule, DxValidatorModule } from 'devextreme-angular';
import { CommonModule } from '@angular/common';

@Component({
  selector: 'ballware-edit-datetime',
  templateUrl: './datetime.component.html',
  styleUrls: ['./datetime.component.scss'],
  imports: [CommonModule, DxDateBoxModule, DxValidatorModule],
  standalone: true
})
export class EditLayoutDatetimeComponent extends WithVisible(WithRequired(WithValidation(WithReadonly(WithValue(WithEditItemLifecycle(WithDestroy()), () => (null as unknown) as string|number|Date))))) implements OnInit, EditItemRef {

  @Input() initialLayoutItem?: EditLayoutItem;

  public layoutItem: EditLayoutItem|undefined;

  public type!: DateType;
  public displayFormat!: string;

  constructor(
    @Inject(TRANSLATOR) private translator: Translator, 
    @Inject(EDIT_SERVICE) private editService: EditService) {
    super();
  }

  ngOnInit(): void {
    if (this.initialLayoutItem) {
      this.initLifecycle(this.initialLayoutItem, this.editService, this);

      this.preparedLayoutItem$
        .pipe(takeUntil(this.destroy$))
        .subscribe((layoutItem) => {
          if (layoutItem) {
            this.initValue(layoutItem, this.editService);
            this.initReadonly(layoutItem, this.editService);
            this.initValidation(layoutItem, this.editService);
            this.initRequired(layoutItem, this.editService);
            this.initVisible(layoutItem);

            this.layoutItem = layoutItem;
            this.type = layoutItem.type as DateType;

            switch (layoutItem.type) {              
              case 'datetime':
                this.displayFormat = this.translator('format.datetime');
                break;
              case 'date':
              default:
                this.displayFormat = this.translator('format.date');
            }
          }
        });
    }
  }

  public getOption(option: string): any {
    switch (option) {
      case 'value':
        return this.value;
      case 'required':
        return this.required$.getValue();
      case 'readonly':
        return this.readonly$.getValue();
      case 'visible':
        return this.visible$.getValue();    
      default:
        throw new Error(`Unsupported option <${option}>`);              
    }
  }

  public setOption(option: string, value: unknown) {
    switch (option) {
      case 'value':
        this.setValueWithoutNotification(value as Date);
        break;
      case 'required':
        this.setRequired(value as boolean);
        break;
      case 'readonly':
        this.setReadonly(value as boolean)
        break;
      case 'visible':
        this.setVisible(value as boolean);
        break;      
      default:
        throw new Error(`Unsupported option <${option}>`);            
    }
  }
}
