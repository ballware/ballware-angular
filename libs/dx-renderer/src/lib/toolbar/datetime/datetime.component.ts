import { AfterViewInit, Component, Input, OnDestroy, ViewChild } from '@angular/core';
import { PageToolbarItem } from '@ballware/meta-model';
import { PageService, ToolbarItemRef } from '@ballware/meta-services';
import { I18NextPipe } from 'angular-i18next';
import { DxDateBoxComponent } from 'devextreme-angular';
import { ValueChangedEvent } from 'devextreme/ui/date_box';

@Component({
  selector: 'ballware-toolbar-datetime',
  templateUrl: './datetime.component.html',
  styleUrls: ['./datetime.component.scss']
})
export class ToolbarDatetimeComponent implements OnDestroy, AfterViewInit {

  @Input() public item?: PageToolbarItem;

  @ViewChild('dateBox', { static: false }) dateBox?: DxDateBoxComponent;

  constructor(private pageService: PageService, private translationService: I18NextPipe) { }

  ngOnDestroy(): void {
      if (this.item?.name) {
        this.pageService.paramEditorDestroyed(this.item.name);
      }
  }

  ngAfterViewInit(): void {

    if (this.item?.name) {
      const toolbarItemRef = {
        getOption: (option) => this.dateBox?.instance.option(option),
        setOption: (option, value) => this.dateBox?.instance.option(option, value)
      } as ToolbarItemRef;

      this.pageService.paramEditorInitialized({ name: this.item.name, item: toolbarItemRef });
    }
  }

  onValueChanged(e: ValueChangedEvent) {
    if (this.item?.name) {
      this.pageService.paramEditorValueChanged({ name: this.item.name, value: e.value });
    }
  }

  getDisplayFormat() {
    return this.translationService.transform('format.datetime');
  }

}
