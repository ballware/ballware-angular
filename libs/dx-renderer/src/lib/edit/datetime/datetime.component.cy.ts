import { MountConfig, mount } from 'cypress/angular';
import { EditLayoutDatetimeComponent } from './datetime.component';

describe(EditLayoutDatetimeComponent.name, () => {
  const config: MountConfig<EditLayoutDatetimeComponent> = {
    declarations: [],
    imports: [],
    providers: []
  }

  it('renders', () => {
     mount(EditLayoutDatetimeComponent, config);
  })
})
