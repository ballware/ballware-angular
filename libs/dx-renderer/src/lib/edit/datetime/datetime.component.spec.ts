import { ComponentFixture, TestBed } from '@angular/core/testing';
import { when } from 'jest-when';

import { EditLayoutDatetimeComponent } from './datetime.component';
import { Provider } from '@angular/core';
import { EDIT_SERVICE, TRANSLATOR } from '@ballware/meta-services';
import { EditLayoutItem } from '@ballware/meta-model';
import { mockedEditServiceContext } from '../../../test/editservice.spec';

describe('EditLayoutDatetimeComponent', () => {
  let component: EditLayoutDatetimeComponent;
  let fixture: ComponentFixture<EditLayoutDatetimeComponent>;

  const mockedTranslator = jest.fn();
  const mockedEditService = mockedEditServiceContext();

  when(mockedTranslator).calledWith('format.date').mockReturnValue('MM/dd/yyyy');
  when(mockedTranslator).calledWith('format.datetime').mockReturnValue('MM/dd/yyyy HH:mm:ss');
        
  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ EditLayoutDatetimeComponent ],
      providers: [        
        {
          provide: TRANSLATOR,
          useValue: mockedTranslator
        },
        {
          provide: EDIT_SERVICE,
          useFactory: () => mockedEditService.mock.object()
        } as Provider         
      ]
    })
    .compileComponents();
  });

  it('should create date', () => {
    fixture = TestBed.createComponent(EditLayoutDatetimeComponent);

    const layoutItem = {
      type: 'date',
      options: {
          dataMember: 'mockedmember',
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    fixture.componentRef.setInput('initialLayoutItem', layoutItem);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should create datetime', () => {
    fixture = TestBed.createComponent(EditLayoutDatetimeComponent);

    const layoutItem = {
      type: 'datetime',
      options: {
          dataMember: 'mockedmember',
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    fixture.componentRef.setInput('initialLayoutItem', layoutItem);
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should apply options', () => {
    fixture = TestBed.createComponent(EditLayoutDatetimeComponent);

    const layoutItem = {
      type: 'datetime',
      options: {
          dataMember: 'mockedmember',
          required: false,
          readonly: false,            
          visible: false
      }
    } as EditLayoutItem;
    
    component = fixture.componentInstance;   
    expect(component).toBeTruthy();

    fixture.componentRef.setInput('initialLayoutItem', layoutItem);
    fixture.detectChanges();

    expect(component.livecycle.getOption('value')).toBe(null);
    expect(component.livecycle.getOption('required')).toBe(false);
    expect(component.livecycle.getOption('readonly')).toBe(false);
    expect(component.livecycle.getOption('visible')).toBe(false);

    component.livecycle.setOption('value', Date.parse('2024-01-01T13:00:00'));
    expect(component.livecycle.getOption('value')).toBe(Date.parse('2024-01-01T13:00:00'));

    component.livecycle.setOption('required', true);
    expect(component.livecycle.getOption('required')).toBe(true);

    component.livecycle.setOption('readonly', true);
    expect(component.livecycle.getOption('readonly')).toBe(true);

    component.livecycle.setOption('visible', true);
    expect(component.livecycle.getOption('visible')).toBe(true);

    expect(() => component.livecycle.getOption('undefined')).toThrowError('Unsupported option <undefined>');
    expect(() => component.livecycle.setOption('undefined', 'any value')).toThrowError('Unsupported option <undefined>');
  });
});
