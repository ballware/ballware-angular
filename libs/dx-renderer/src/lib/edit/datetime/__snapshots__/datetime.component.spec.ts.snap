// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLayoutDatetimeComponent should create date 1`] = `
<ballware-edit-datetime
  currentValue$={[Function BehaviorSubject]}
  destroy$={[Function Subject]}
  displayFormat={[Function String]}
  editService={[Function ]}
  editorEntered$={[Function Subject]}
  editorEvent$={[Function Subject]}
  emailValidation$={[Function BehaviorSubject]}
  globalReadonly="false"
  initialLayoutItem={[Function Object]}
  layoutItem={[Function Object]}
  notifyValueChange$={[Function Subject]}
  onEntered={[Function Function]}
  onEvent={[Function Function]}
  preparedLayoutItem$={[Function BehaviorSubject]}
  readonly$={[Function BehaviorSubject]}
  refreshValueTrigger$={[Function Subject]}
  required$={[Function BehaviorSubject]}
  requiredValidation$={[Function BehaviorSubject]}
  translator={[Function Function]}
  type={[Function String]}
  validationRules$={[Function Observable]}
  visible$={[Function BehaviorSubject]}
>
  <dx-date-box
    class="mt-2 dx-validator dx-visibility-change-handler dx-show-invalid-badge dx-datebox dx-textbox dx-texteditor dx-show-clear-button dx-dropdowneditor-button-visible dx-editor-outlined dx-texteditor-empty dx-widget dx-auto-width dx-dropdowneditor dx-datebox-date dx-datebox-calendar"
  >
    <div
      class="dx-dropdowneditor-input-wrapper"
    >
      <input
        name="mockedmember"
        type="hidden"
        value=""
      />
      <div
        class="dx-texteditor-container"
      >
        <div
          class="dx-texteditor-input-container"
        >
          <input
            aria-autocomplete="none"
            aria-expanded="false"
            aria-haspopup="true"
            autocomplete="off"
            class="dx-texteditor-input"
            role="combobox"
            tabindex="0"
            type="text"
          />
          <div
            class="dx-placeholder"
            data-dx_placeholder=""
          />
        </div>
        <div
          class="dx-texteditor-buttons-container"
        >
          <span
            class="dx-clear-button-area"
          >
            <span
              class="dx-icon dx-icon-clear"
            />
          </span>
          <div
            aria-label="Select"
            class="dx-widget dx-button-normal dx-dropdowneditor-button"
            role="button"
          >
            <div
              class="dx-button-content"
            >
              <div
                class="dx-dropdowneditor-icon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </dx-date-box>
</ballware-edit-datetime>
`;

exports[`EditLayoutDatetimeComponent should create datetime 1`] = `
<ballware-edit-datetime
  currentValue$={[Function BehaviorSubject]}
  destroy$={[Function Subject]}
  displayFormat={[Function String]}
  editService={[Function ]}
  editorEntered$={[Function Subject]}
  editorEvent$={[Function Subject]}
  emailValidation$={[Function BehaviorSubject]}
  globalReadonly="false"
  initialLayoutItem={[Function Object]}
  layoutItem={[Function Object]}
  notifyValueChange$={[Function Subject]}
  onEntered={[Function Function]}
  onEvent={[Function Function]}
  preparedLayoutItem$={[Function BehaviorSubject]}
  readonly$={[Function BehaviorSubject]}
  refreshValueTrigger$={[Function Subject]}
  required$={[Function BehaviorSubject]}
  requiredValidation$={[Function BehaviorSubject]}
  translator={[Function Function]}
  type={[Function String]}
  validationRules$={[Function Observable]}
  visible$={[Function BehaviorSubject]}
>
  <dx-date-box
    class="mt-2 dx-validator dx-visibility-change-handler dx-show-invalid-badge dx-datebox dx-textbox dx-texteditor dx-show-clear-button dx-dropdowneditor-button-visible dx-editor-outlined dx-texteditor-empty dx-widget dx-auto-width dx-dropdowneditor dx-datebox-datetime dx-datebox-calendar"
  >
    <div
      class="dx-dropdowneditor-input-wrapper"
    >
      <input
        name="mockedmember"
        type="hidden"
        value=""
      />
      <div
        class="dx-texteditor-container"
      >
        <div
          class="dx-texteditor-input-container"
        >
          <input
            aria-autocomplete="none"
            aria-expanded="false"
            aria-haspopup="true"
            autocomplete="off"
            class="dx-texteditor-input"
            role="combobox"
            tabindex="0"
            type="text"
          />
          <div
            class="dx-placeholder"
            data-dx_placeholder=""
          />
        </div>
        <div
          class="dx-texteditor-buttons-container"
        >
          <span
            class="dx-clear-button-area"
          >
            <span
              class="dx-icon dx-icon-clear"
            />
          </span>
          <div
            aria-label="Select"
            class="dx-widget dx-button-normal dx-dropdowneditor-button"
            role="button"
          >
            <div
              class="dx-button-content"
            >
              <div
                class="dx-dropdowneditor-icon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </dx-date-box>
</ballware-edit-datetime>
`;
